### PR TITLE
Add median `val` monitoring tool

### PR DESCRIPTION
**Pull request by teghnet**
This actually includes #62

### CODE REVIEW

File: pkg/log/grafana/log.go [hunk 1]
```diff
@@ -103,19 +101,20 @@ func New(ctx context.Context, level log.Level, cfg Config) (log.Logger, error) {
 		level:  level,
 		fields: log.Fields{},
 	}
-	go l.pushRoutine()
 	return l, nil
 }
 
 type logger struct {
+	ctx    context.Context
+	waitCh chan error
```
Comment (MDobak) [inline pkg/log/grafana/log.go:109]: These two fields should be part of `shared`. There is no need to copy them every time a new field is added.

File: pkg/log/grafana/log.go [hunk 2]
```diff
@@ -349,8 +353,26 @@ func (c *logger) addMetricPoint(m Metric, mk metricKey, mv metricValue) {
...
+	return nil
+}
+
+func (c *logger) Wait() chan error {
+	return c.waitCh
+}
+
 // pushRoutine pushes metrics in interval defined in c.interval.
 func (c *logger) pushRoutine() {
+	defer c.logger.Info("Stopped")
```
Comment (MDobak) [inline pkg/log/grafana/log.go:374]: If there is a `Stopped` log, there should also be a `Starting` log.

File: pkg/log/grafana/log_test.go [hunk 3]
```diff
@@ -70,6 +71,30 @@ func TestLogger(t *testing.T) {
...
+			metrics: []Metric{
+				{MatchMessage: regexp.MustCompile("foo"), Name: "test", Value: "key", ParserFunc: loggerConfig.ToFloatVersion},
+			},
+			want: []want{
+				{name: "test", value: -1002003},
+			},
+			logs: func(l log.Logger) {
+				l.WithField("key", "1.2.3-test.0").Info("foo")
+			},
+		},
```
Comment (MDobak) [inline pkg/log/grafana/log_test.go:97]: I think tests for `ToFloatVersion` should be moved to the `config` pkg as well.

File: pkg/log/grafana/log.go [hunk 4]
```diff
@@ -387,6 +409,11 @@ func (c *logger) pushMetrics() {
 	// Grafana is complete.
 	once.Do(c.mu.Unlock)
 	reqBody, err := json.Marshal(metrics)
+	defer func() {
+		if err != nil {
+			c.waitCh <- err
+		}
+	}()
```
Comment (MDobak) [inline pkg/log/grafana/log.go:416]: I'm not sure if this is correct behavior. An error returned to `waitCh` indicates that there was an error that caused a service to stop or an error that happened during stopping a service. In this case, this error is logged anyway a few lines below and it didn't match the mentioned conditions.

File: pkg/log/grafana/log.go [hunk 5]
```diff
@@ -349,8 +353,26 @@ func (c *logger) addMetricPoint(m Metric, mk metricKey, mv metricValue) {
 		Debug("New metric point")
 }
 
+var _ log.LoggerService = (*logger)(nil)
+
+func (c *logger) Start(ctx context.Context) error {
+	if ctx == nil {
+		return fmt.Errorf("context is nil")
+	}
```
Comment (MDobak) [inline pkg/log/grafana/log.go:361]: There is a missing check against for double starting of the service.

File: pkg/config/logger/logger.go [hunk 6]
```diff
@@ -30,9 +31,7 @@ import (
 	"github.com/chronicleprotocol/oracle-suite/pkg/log/grafana"
 )
 
-var grafanaLoggerFactory = func(lvl log.Level, cfg grafana.Config) (log.Logger, error) {
-	return grafana.New(context.Background(), lvl, cfg)
-}
+var grafanaLoggerFactory = grafana.New
```
Comment (MDobak) [inline pkg/config/logger/logger.go:34]: 👍

File: pkg/config/logger/logger.go [hunk 7]
```diff
@@ -52,6 +51,7 @@ type grafanaLogger struct {
 }
 
 type grafanaMetric struct {
+	Type         string              `json:"type"`
```
Comment (MDobak) [inline pkg/config/logger/logger.go:54]: Missing documentation in `README`.

**Review by MDobak — CHANGES_REQUESTED** [review]

File: pkg/log/grafana/log_test.go (hunk 3 above)
Comment (teghnet) [reply pkg/log/grafana/log_test.go]: Removed it completely.

File: pkg/log/chain/log.go [hunk 8]
```diff
@@ -32,6 +34,32 @@ type logger struct {
...
+	}
+	return nil
+}
+
+func (c *logger) Wait() chan error {
+	ch := make(chan error)
+	for _, l := range c.loggers {
+		if srv, ok := l.(supervisor.Service); ok {
+			go func() {
+				ch <- <-srv.Wait()
```
[the remark this answers is not in the document]
Comment (teghnet) [reply pkg/log/chain/log.go:55]: The items will already be log.Loggers, but this limits dependencies, so yeah. :+1:

And, yes not closing chennel.

File: pkg/log/grafana/log.go (hunk 1 above)
Comment (teghnet) [reply pkg/log/grafana/log.go:109]: ok

File: pkg/log/grafana/log.go (hunk 2 above)
Comment (teghnet) [reply pkg/log/grafana/log.go:374]: ok

File: pkg/log/grafana/log.go (hunk 5 above)
Comment (teghnet) [reply pkg/log/grafana/log.go:361]: ok

File: pkg/log/grafana/log.go (hunk 4 above)
Comment (teghnet) [reply pkg/log/grafana/log.go:420]: bueno

File: pkg/log/grafana/log.go (hunk 2 above)
Comment (teghnet) [reply pkg/log/grafana/log.go:378]: ok

**Pull request merged**